packages/backend/src/apps/lettersg/actions/create-letter/schema.ts

```diff
@@ -0,0 +1,67 @@
+import { DateTime } from 'luxon'
+import { z } from 'zod'
+
+export const requestSchema = z
+  .object({
+    templateId: z.string().trim().min(1, {
+      message: 'Please do not leave the template id empty',
+    }),
+    letterParams: z
+      .array(
+        z.object({
+          field: z
+            .string()
+            .trim()
+            .min(1, 'Please do not leave the field empty')
+            .nullish(),
+          value: z
+            .string()
+            .trim()
+            .min(1, 'Please do not leave the value empty')
+            .nullish(),
+        }),
+      )
+      .transform((params, context) => {
+        const result: Record<string, string> = Object.create(null)
+        for (const param of params) {
+          // no null fields or values are allowed
+          if (!param.field) {
+            context.addIssue({
+              code: z.ZodIssueCode.custom,
+              message: 'Please do not leave the field empty',
+            })
+            return z.NEVER
+          }
+          if (!param.value) {
+            context.addIssue({
+              code: z.ZodIssueCode.custom,
+              message: 'Please do not leave the value empty',
+            })
+            return z.NEVER
+          }
+          result[param.field] = param.value
+        }
+        return result
+      })
+      .nullish(),
+  })
+  .transform((data) => ({
+    templateId: Number(data.templateId),
+    letterParams: data.letterParams,
+  }))
+
+export const responseSchema = z
+  .object({
+    publicId: z.string().nullish(),
```

Comment:
actually i think ok to remove nullish if the response is useless without it.

Thinking is that we don't want the pipe to silently succeed

Reply:
ok got it, thanks!